packages/common/src/graphql/entities.ts

```diff
@@ -1,14 +1,108 @@
 // Copyright 2020-2021 OnFinality Limited authors & contributors
 // SPDX-License-Identifier: Apache-2.0
 
-import {GraphQLSchema, GraphQLObjectType, isObjectType} from 'graphql';
+import assert from 'assert';
+import {
+  assertListType,
+  getDirectiveValues,
+  getNullableType,
+  GraphQLField,
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLSchema,
+  isListType,
+  isNonNullType,
+  isObjectType,
+} from 'graphql';
 import {DirectiveName} from './constant';
 
-export function getAllEntities(schema: GraphQLSchema): GraphQLObjectType[] {
-  return Object.entries(schema.getTypeMap())
-    .filter(([, node]) => node.astNode?.directives?.find(({name: {value}}) => value === DirectiveName.Entity))
-    .map(([, node]) => node)
+import {
+  FieldScalar,
+  GraphQLEntityField,
+  GraphQLModelsRelations,
+  GraphQLModelsType,
+  GraphQLRelationsType,
+} from './types';
+
+export function getAllEntitiesRelations(schema: GraphQLSchema): GraphQLModelsRelations {
+  const entities = Object.values(schema.getTypeMap())
+    .filter((node) => node.astNode?.directives?.find(({name: {value}}) => value === DirectiveName.Entity))
+    .map((node) => node)
     .filter(isObjectType);
+
+  const entityNameSet = entities.map(function (entity) {
+    return entity.name;
+  });
+
+  const modelRelations = {models: [], relations: []} as GraphQLModelsRelations;
+  const derivedFrom = schema.getDirective('derivedFrom');
+  for (const entity of entities) {
+    const newModel: GraphQLModelsType = {
+      name: entity.name,
+      fields: [],
+    };
+    for (const field of Object.values(entity.getFields())) {
+      const typeString = extractType(field.type).toString();
+      const directValues = getDirectiveValues(derivedFrom, field.astNode);
+      //If is a basic scalar type
+      if (Object.values(FieldScalar).includes(typeString)) {
+        newModel.fields.push(packEntityField(typeString, field, false));
+      }
+      // If is a foreign key
+      else if (entityNameSet.includes(typeString) && !directValues) {
+        newModel.fields.push(packEntityField(typeString, field, true));
+        modelRelations.relations.push({
+          from: entity.name,
+          type: 'belongsTo',
+          to: typeString,
+          foreignKey: `${field.name}Id`,
+        } as GraphQLRelationsType);
+      }
+      // If is derivedFrom
+      else if (entityNameSet.includes(typeString) && directValues) {
+        modelRelations.relations.push({
+          from: entity.name,
+          type: isListType(isNonNullType(field.type) ? getNullableType(field.type) : field.type) ? 'hasMany' : 'hasOne',
+          to: typeString,
+          foreignKey: `${directValues.field}Id`,
+          fieldName: field.name,
+        } as GraphQLRelationsType);
+      } else {
+        throw new Error(`${typeString} is not an valid type`);
+      }
+    }
+    modelRelations.models.push(newModel);
+  }
+  validateRelations(modelRelations);
```

Comment:
`validateRelations` looks like a test function, we need to make sure every model/relations is valid before insert;
 I don't think we need this validate here.

Reply:
To validate relations will require prepare full set of models in advance. Actually, only relations with 'hasMany' need to be validated here, and we need to check it has valid foreignKey in the corresponding entity. 

Also, if validation failed, it will throw errors before it gets inserted into the database.